README.md
Asker [![NPM version][npm-image]][npm-link] [![Build status][build-image]][build-link]
=====

[![Dependency status][deps-image]][deps-link]
[![devDependency status][devdeps-image]][devdeps-link]

Asker is a wrapper for `http.request` method, which incorporates:
* response deflating using gzip,
* requests retrying,
* connection pools tuning.

If you are looking for a module to fetch 3rd-party web content (pages, RSS, files or something else), don't waste your time and look at the [request](http://npm.im/request) module, because `asker` doesn't support cookies and redirects out of the box.

`Asker`'s main goal is to communicate between frontends and backends that use some kind of [SLA](http://en.wikipedia.org/wiki/Service-level_agreement).

## Quick start

```javascript
var ask = require('asker');

ask({ host : 'ya.ru' }, function(error, response) {
    if (error) {
        return error.log();
    }

    console.log('Response retrieved in ' + response.meta.time.total + 'ms');
    console.log('==========\n', response.data, '\n==========');
});
```

## Options

All parameters are optional.

* `{String} host="localhost"`
* `{String} hostname` — the same as the `host`, used for compatibility with results of the `url.parse()`. Have higher priority than the `host`.
* `{Number} port=80`
* `{String} path="/"`
* `{String} protocol="http:"` — `http:` or `https:`
* `{String} url` — Shorthand alternative for `protocol`, `hostname`, `port` and `path` options
* `{String} method="GET"`
* `{Object} headers` — HTTP headers
* `{Object} query` — Query params
* `{String} requestId=""` — Request ID, used in log messages
* `{*} body` — request body. If it's an `Object` — `JSON.stringify` is applied, otherwise it's converted to `String`.
* `{String} bodyEncoding="string"` — Body encoding method (`string`, `json`, `urlencoded`, `multipart` or self implemented). [More info](#body-encoding).
* `{Number} maxRetries=0` — Max number of retries allowed for the request
* `{Function} onretry(reason Error, retryCount Number)` — called when retry happens. By default it does nothing. As an example, you can pass a function that logs a warning.
* `{Number} timeout=500` — timeout from the moment, when a socket was given by a pool manager.
* `{Number} queueTimeout=timeout+50` — timeout from the moment, when asker initiated the request. Useful if pool manager failed to provide a socket for any reason.
* `{Boolean} allowGzip=true` — allows response compression with gzip
* `{Function} statusFilter` — status codes processing, see [Response status codes processing](#response-status-codes-processing) section for details.
* `{Object|false} agent` — http.Agent options, see [Connection pools tuning](#connection-pools-tuning) section for details.

## Response format

Succesful requests will return data and additional information in the following format:

`{Object} response`
* `{*} data` received data (if response body wasn't provided, `null` is returned)
* `{Number} statusCode` http status code
* `{Object} headers` returned http headers (names are lowercased)
* `{Object} meta` meta information
    * `{Object} time` request timers
        * `{Number} network` from socket open until the request completion
        * `{Number} total` total execution time
    * `{Object} options` options that you provided when created an `Asker` request
    * `{Object} retries`
        * `{Number} used` number of retries used
        * `{Number} limit` retries limit for a given request

## Response status codes processing

When response status code is received, `asker` passes status code through the filter function, which should determine whether this response code is acceptable. And if not, whether is it necessary to retry a request.

The only filter function argument is `code`:
* `{Number} code` is a response status code provided by `asker`.

Function must return an Object with two fields:
* `{Boolean} accept` — whether to accept response with a given status code;
* `{Boolean} isRetryAllowed` — whether to retry an unaccepted request.

Result must be returned ASAP, because filter's execution time WILL affect request timeouts.

Default filter accepts codes `200` and `201` and allows retries for all codes except `400-499`.

Let's make a quick example. Suppose, we want to accept only responses with `200`, `201` and `304` status codes and do not want to retry requests for `4xx`.

```javascript
var ask = require('asker');

function filter(code) {
    return {
        accept : ~[200, 201, 304].indexOf(code),
        isRetryAllowed : 400 > code || code > 499
    }
}

ask({ host: 'data-feed.local', statusFilter : filter }, function(error, response) {
    // @see http://npm.im/terror for details about error codes
    if (error.code === ask.Error.CODES.UNEXPECTED_STATUS_CODE) {
        console.log('Response status code is not 200, 201 or 304');
    }

    // ...
});
```

## Body encoding

Body encoder converts `body` to corresponding format and sets `Content-type` header.

### Built-in encoders

* `string` — Used by default. Converts `body` to `String`. Accepts all types.
* `json` — Applies `JSON.stringify` to the `body`. Accepts all types.
* `urlencoded` — Converts `body` to query string. Accepts `Object`.
* `multipart` — Formats `body` according to [multipart/form-data spec](http://www.w3.org/TR/html4/interact/forms.html#h-17.13.4.2). Accepts `Object` (or `Buffer` object).
* `raw` – Use body as is. Accepts instance of Buffer. Remember to set `content-type` header manually if required.

### Content-type

If you pass `Buffer` as property value, mime-type `application/octet-stream` will be applied. And property name will be used as file name.

Otherwise, you can pass additional info (mime-type and filename) in parameter's description:
```javascript
ask({
    bodyEncoding : 'multipart', // encoder name

    body : {
        'sample.mp3' : buffer, // an instance of Buffer, "sample.mp3" will be used as file name

        image : {
            filename : 'image.jpg',
            mime : 'image/jpeg',
            data : image_buffer // an instance of Buffer
        }
    }
}, function(error, response) {
    /* ... */
});
```

If you need to send multiple files in single field wrap them in an array.
```javascript
ask({
    bodyEncoding : 'multipart',

    body : {
        images : [
            'sample.mp3' : buffer,
            {
                filename : 'pic.jpg',
                mime : 'image/jpeg',
                data : pic_buffer
            }
        ]
    }
}, function(error, response) {
    /* ... */
});
```

### Exceptions

If you use body encoder, Asker can throw following errors:

* `BODY_ENCODER_NOT_EXISTS` – unknown `bodyEncoder` has been passed;
* `BODY_INCORRECT_TYPE` – `body`'s type is not allowed by the encoder.

### Custom encoders

To implement you own body encoder, you must add an encoding function as the `Asker.bodyEncoders` property. Property name will be used as the encoder name.

Example:
```javascript
var Asker = require('asker');

// encoder name is 'trimText'
Asker.bodyEncoders.trimText = function(body, setContentType) {
    // throw error if passed body format is not acceptable for your encoder
    if (['number', 'string', 'boolean'].indexOf(typeof body) === -1) {
        throw AskerError.createError(AskerError.CODES.UNEXPECTED_BODY_TYPE, {
            type : typeof body,
            expectedTypes : 'Object'
        });
    }

    // 'content-type' header will be set to 'text/plain'
    setContentType('text/plain');

    return String(data).trim();
};
```

Note: `setContentType` sets `Content-Type` header only if header was not set before. But you can force overriding by passing `true` as second argument:

```javascript
Asker.bodyEncoders.trimText = function(body, setContentType) {
    setContentType('nyan/colorful', true);

    return 'Colorful nyan cat';
};
```

## Connection pools tuning

### The problem

At this point of time (version `0.10` and below) node.js provides a socket pool manager that works as follows:

* by default `globalAgent` is used for all outgoing http requests;
* each `Agent` instance, including `globalAgent`, has a `maxSockets` property, which you can change;
* socket limit is set for each unique __host-port pair__, that is served by this particular `Agent`.

That is sometimes an unwanted behaviour. Let's take an example.

You have two backends: `backend:3000` and `backend:4000`. First backend is indispensable for the application, but the second one is complementary. E.g., it makes http calls for the advertisements that you show later.

Under heavy load this additional backend (which is usually less fault-tolerant) may occupy all sockets that OS provides for the whole `node` process, because default pool manager cares only about __host-port__ pairs. `defaultAgent` does not anyhow correct each backend' socket limit according to process limit.

### Solution

How `Asker` can help you manage this problem? `Asker` can create a custom instance of `http.Agent` for any given backend. And you can set up a `maxSockets` property by calculating each backend priority.

### API

`{Object} agent` `http.Agent` options:

* `{String} name='globalAgent'` unique name for this backend
* `{Number} maxSockets=1024` socket limit for this backend
* `{Boolean} persistent=true` either agent would be deleted when it's last socket is removed. You can consider setting it to false, if you create agent's `name` in runtime.

## Error handling

Asker produces errors using [Terror](http://npm.im/terror), so you can setup your own logger and use `error.log()` method for logging.

If you already use Terror and created a logger for Terror itself, you shouldn't setup it again for AskerError.

`AskerError` class is available via `request('asker').Error` property. So you can, for example, localize error messages or customize it in your own way.
[npm-image]: https://img.shields.io/npm/v/asker.svg?style=flat
[npm-link]: https://npmjs.org/package/asker

CHANGELOG
asker (0.6.2); urgency=low

  * allow an instance of Agent as a value of the `agent` option.

 -- Phillip Kovalev <twilightfeel@gmail.com>  Mon, 18 May 2015 19:36:03 +0300

asker (0.6.1); urgency=low

  * fill 'execution' timer on exceeded retries (by Dmitry Sorin https://github.com/1999)

 -- Phillip Kovalev <twilightfeel@gmail.com>  Mon, 30 Mar 2015 11:34:50 +0300

asker (0.6.0); urgency=low

  * updates terror module version to 1.0.0

 -- Phillip Kovalev <twilightfeel@gmail.com>  Mon, 08 Dec 2014 10:03:42 +0300

asker (0.5.0); urgency=low

  * implements `hostname` option which is the same as `host` and introduced
    to support results of `url.parse()` as options for asker.
    `hostname` has higher priority than `host` if both passed.
  * updates terror module version to 0.4.x

 -- Phillip Kovalev <twilightfeel@gmail.com>  Fri, 03 Oct 2014 16:10:12 +0300

asker (0.4.6); urgency=low

  * implement encoding of multiple files for single parameter by multipart encoder

 -- Phillip Kovalev <twilightfeel@gmail.com>  Thu, 25 Jun 2014 17:23:01 +0400

asker (0.4.5); urgency=low

  * add support for `false` value for the option `agent`

 -- Phillip Kovalev <twilightfeel@gmail.com>  Tue, 13 May 2014 15:22:37 +0400

asker (0.4.4); urgency=low

  * fix broken timings in the error messages on failed retries

 -- Phillip Kovalev <twilightfeel@gmail.com>  Tue, 13 May 2014 15:22:37 +0400

asker (0.4.3); urgency=low

  * use `requestId` to form the contimer timers IDs

 -- Phillip Kovalev <twilightfeel@gmail.com>  Mon, 12 May 2014 18:40:20 +0400

asker (0.4.2); urgency=low

  * use contimer module to measure requests execution and network time

 -- Phillip Kovalev <twilightfeel@gmail.com>  Mon, 12 May 2014 13:40:20 +0400

asker (0.4.1); urgency=low

  * return 'Request.DEFAULT_OPTIONS.port' which was lost on https merge

 -- Phillip Kovalev <twilightfeel@gmail.com>  Tue, 08 Apr 2014 10:00:22 +0400

asker (0.4.0); urgency=low

  * add some support for https (be careful with agents usage for https)

 -- Phillip Kovalev <twilightfeel@gmail.com>  Mon, 07 Apr 2014 14:55:13 +0400

asker (0.3.2); urgency=low

  * fix serialization of non-string literals in multipart encoder:
    all non-string literals automatically casted as strings

 -- Phillip Kovalev <twilightfeel@gmail.com>  Fri, 06 Dec 2013 16:13:40 +0400

asker (0.3.1); urgency=low

  * allow to list options accepted by asker request constructor

 -- Phillip Kovalev <twilightfeel@gmail.com>  Fri, 1 Nov 2013 21:30:00 +0400

asker (0.3.0); urgency=low

  * add "raw" body encoder to pass raw Buffer as request body

 -- Phillip Kovalev <twilightfeel@gmail.com>  Thu, 24 Oct 2013 21:23:05 +0400

asker (0.2.2); urgency=low

  * fix error name duplication in AskerError
  * remove redundant UNKNOWN_ERROR code overriding

 -- Phillip Kovalev <twilightfeel@gmail.com>  Fri, 23 Aug 2013 13:50:35 +0400

asker (0.2.1); urgency=low

  * change `terror` dependency version to 0.3.x

 -- Phillip Kovalev <twilightfeel@gmail.com>  Sun, 11 Aug 2013 19:21:31 +0400

asker (0.2.0); urgency=low

  * implement body encoding and support for custom encoders
  https://github.com/nodules/asker/issues/9

  * allow to pass callback to Request#execute method
  https://github.com/nodules/asker/issues/59

  * add jscs to npm test toolchain

  * response.data must be a buffer, not a string
  https://github.com/nodules/asker/issues/6

 -- Phillip Kovalev <twilightfeel@gmail.com>  Fri, 02 Aug 2013 19:09:27 +0400

asker (0.1.13); urgency=low

  * fulfill error message for code UNEXPECTED_STATUS_CODE
  https://github.com/nodules/asker/issues/72

 -- Phillip Kovalev <twilightfeel@gmail.com>  Fri, 26 Jun 2013 11:07:15 +0400

asker (0.1.12); urgency=critical

  * full code coverage by tests
  https://github.com/nodules/asker/issues?labels=test&milestone=3&page=1&state=closed

  * drop support for Node.js < 0.8.21 due to the bugs in the "http" module

  * fix incorrect behaviour of Request#isRunning (always returns `false`)
  https://github.com/nodules/asker/issues/44

  * constructor compiles request body always when `options.body`
  have defined
  https://github.com/nodules/asker/issues/1

  * fix redundant callbacks is the case when "content-length" header is less
  than actual response body length
  https://github.com/nodules/asker/issues/46

  * fix missing callback call in the case when "content-length" header is
  more than actual response body length
  https://github.com/nodules/asker/issues/47

  * move `queueTimeout` default value calculation from request excution time
  to constructor
  https://github.com/nodules/asker/issues/48

  * callback has became optional argument of the request constructor
  https://github.com/nodules/asker/issues/54

  * fix network timer resolving
  https://github.com/nodules/asker/issues/58

  * request can be tried to execute with a dead agent
  https://github.com/nodules/asker/issues/60


lib/asker.js
var http = require('http'),
    https = require('https'),
    zlib = require('zlib'),
    url = require('url'),
    extend = require('extend'),
    contimer = require('contimer'),
    agentFactory = require('./advanced_agent'),
    AskerError = require('./error');

/**
 * Shorthand to call Object#hasOwnProperty in context of the obj with propName argument.
 * @param {Object} obj context object to check
 * @param {String} propName property name to check
 * @returns {Boolean}
 */
function has (obj, propName) {
    return Object.prototype.hasOwnProperty.call(obj, propName);
}

/**
 * @typedef {Object} AgentOptions
 * @property {String}  [name=globalAgent] Agent name, use 'globalAgent' for http.globalAgent
 * @property {Number}  [maxSockets=1024] Pool size, used only if new agent defined
 * @property {Boolean} [persistent=true] Non-persistent agents removed when queue empty
 */

/**
 * @typedef  {Object} RequestOptions
 * @property {String}   [url] shorthand alternative for host, port and path options
 * @property {String}   [host=localhost]
 * @property {Number}   [port=80]
 * @property {String}   [protocol=http:] 'http:' or 'https:'
 * @property {String}   [path=/]
 * @property {String}   [method=GET] HTTP-method
 * @property {Object}   [headers] HTTP headers hash
 * @property {Object}   [query] Query params hash
 * @property {String}   [requestId=''] Request identifier for error messages
 * @property {*}        [body] Request body
 * @property {String}   [bodyEncoding] Body encoding method = multipart|urlencoded|text|stringify (default)
 *    if multipart chosen, there are two ways to transfer file content at body:
 *    directly as buffer: { param_name: <Buffer ...> }
 *    with extended info: { param_name: {filename: 'pic.jpg', mime: 'image/jpeg', data: <Buffer ...>} }
 * @property {Number}   [maxRetries=0] Max number of allowed retries for request
 * @property {Function} [onretry] (reason Error, retryCount Number) called on retries
 * @property {Number}   [timeout=500] Socket timeout
 * @property {Number}   [queueTimeout=timeout+50] Queue timeout
 * @property {Boolean}  [allowGzip=true] Allows response compression with gzip
 * @property {Function} [statusFilter] (code Number) Filter which determines acceptable status codes
 *    by default only 200 and 201 codes acceptable and retries allowed for all codes, except range from 400 to 499.
 *    Must returns object { accept : Boolean, isRetryAllowed : Boolean }.
 * @property {AgentOptions} [agent] http.Agent options
 */

/**
 * Create new instance and calls the `execute()` method if called as function (without `new`).
 *
 * @constructor
 * @param {RequestOptions} [options] object with only `host` field required
 * @param {Function} [callback]
 */
function Request(options, callback) {
    var parsedUrl,
        parsedPath,
        acceptEncoding;

    // execute request immediately after construction if called without `new`
    if ( ! (this instanceof Request)) {
        return (new Request(options, callback)).execute();
    }

    // override default options with passed hash
    this.options = extend({}, Request.DEFAULT_OPTIONS, options);

    // set hostname option if not set
    this.options.hostname = this.options.hostname || this.options.host;

    // setup callbacks
    this._callback = callback;
    this._onretry = this.options.onretry;

    // uppercase method name
    this.options.method = this.options.method.toUpperCase();

    // override status codes filter if passed in the options
    if (typeof this.options.statusFilter === 'function') {
        this.statusCodeFilter = this.options.statusFilter;
    }

    // lowercase headers names
    this.options.headers = (options && options.headers) ?
        Object.keys(this.options.headers).reduce(function(headers, headerName) {
            headers[headerName.toLowerCase()] = options.headers[headerName];

            return headers;
        }, {}) :
        {};

    // produce `host`, `port` and `path` options from the `url` option
    if (this.options.url) {
        // allow url without protocol (use 'http' by default)
        if ( ! /^https?:\/\//ig.test(this.options.url)) {
            this.options.url = 'http://' + this.options.url;
        }

        parsedUrl = url.parse(this.options.url, true);

        this.options.protocol = parsedUrl.protocol;
        this.options.host = this.options.hostname = parsedUrl.hostname;
        this.options.port = parseInt(parsedUrl.port, 10);
        this.options.path = parsedUrl.path;
    }

    // set default port (80 for http and 443 for https)
    if ( ! this.options.port) {
        this.options.port = this.options.protocol === 'https:' ? 443 : 80;
    }

    // rebuild path with query params
    // `query` hash properties has higher priority than the specified in the `path` string does
    if (this.options.query) {
        parsedPath = url.parse(this.options.path, true);

        // `search` prop has higher priority than the `query` then remove it
        delete parsedPath.search;
        extend(parsedPath.query, this.options.query);

        this.options.path = url.format(parsedPath);
    }

    // build request body
    if (typeof this.options.body !== 'undefined') {
        this.compileBody();

        if ( ! this.options.headers['content-length']) {
            this.options.headers['content-length'] = this.options.body.length;
        }
    }

    // add "gzip" to the "accept-encoding" header
    if (this.options.allowGzip) {
        acceptEncoding = this.options.headers['accept-encoding'] || '*';

        if (acceptEncoding.indexOf('gzip') === -1) {
            this.options.headers['accept-encoding'] = 'gzip, ' + acceptEncoding;
        }
    }

    // calculate queueTimeout option if not defined
    if (isNaN(this.options.queueTimeout)) {
        this.options.queueTimeout = this.options.timeout + this.QUEUE_TIMEOUT_DELTA;
    }
}

/**
 * expose AskerError for end user
 * @type {Fucntion} Error constructor, inherited from Terror
 * @see http://npm.im/terror
 */
Request.Error = AskerError;

/**
 * default Request options
 * @type {Object}
 */
Request.DEFAULT_OPTIONS = {
    protocol : 'http:',
    host : 'localhost',
    path : '/',
    method : 'GET',
    bodyEncoding : 'string',
    maxRetries : 0,
    timeout : 500,
    allowGzip : true,
    requestId : '',
    url : undefined,
    headers : undefined,
    query : undefined,
    body : undefined,
    onretry : undefined,
    statusFilter : undefined,
    queueTimeout : undefined,
    agent : undefined,
    port : undefined
};

/**
 * Body encoding methods
 * @type {Object}
 * @see ./body_encoders.js
 */
Request.bodyEncoders = require('./body_encoders');

/**
 * Compiles HTTP request body from Request#options.body object
 * using body encoders
 * @throws {AskerError} UNEXPECTED_ENCODER_ERROR
 */
Request.prototype.compileBody = function() {
    var encoderName = this.options.bodyEncoding,
        encoders = this.constructor.bodyEncoders;

    if ( ! has(encoders, encoderName)) {
        throw AskerError.createError(
            AskerError.CODES.BODY_ENCODER_NOT_EXIST, { encoder : encoderName });
    }

    try {
        // execute encoder in the context of the instance of Request
        this.options.body = encoders[encoderName].call(this, this.options.body, this.setContentType.bind(this));
    } catch (encoderError) {
        throw AskerError
            .ensureError(encoderError, AskerError.CODES.UNEXPECTED_ENCODER_ERROR)
            .bind({ encoder : encoderName });
    }
};

/**
 * Set 'content-type' header value
 * if it was not previously set or override argument evals as `true`
 * @param {String} contentType
 * @param {Boolean} [override]
 * @returns {String} actual "Content-Type" header value
 */
Request.prototype.setContentType = function(contentType, override) {
    if (override || ! has(this.options.headers, 'content-type')) {
        this.options.headers['content-type'] = contentType;
    }

    return this.options.headers['content-type'];
};

/**
 * used to calculate queueTimeout = timeout + QUEUE_TIMEOUT_DELTA
 * @type {Number}
 * @const
 */
Request.prototype.QUEUE_TIMEOUT_DELTA = 50;

/**
 * default pool sizes
 */
http.globalAgent.maxSockets = 1024;
https.globalAgent.maxSockets = 1024;
http.Agent.defaultMaxSockets = 1024;
https.Agent.defaultMaxSockets = 1024;

/**
 * Returns existing agent from agents pool or creates new if no one exists with requested name.
 * Method have the side-effect: if you call it twice or more with same agent name,
 * but other options is different, then Agent willn't be reconfigured.
 *
 * @param {Request} request instance
 * @returns {Agent|AdvancedAgent|Boolean}
 */
Request.getAgent = function(request) {
    var agent = request.options.agent;

    if (agent === false) {
        return false;
    }

    // checks custom agent
    if (typeof agent === 'object' && agent.addRequest && agent.createSocket) {
        return agent;
    }

    var protocol = request.options.protocol.split(':')[0],
        name = agent && agent.name,
        AdvancedAgent = agentFactory(protocol),
        assignedAgent;

    // returns `http.globalAgent` if agent name is undefined or equals 'globalAgent'
    if ( ! name || name === 'globalAgent') {
        assignedAgent = protocol === 'https' ? https.globalAgent : http.globalAgent;
    } else {
        assignedAgent = AdvancedAgent.get(name);
    }

    // if agent is not globalAgent or is not exists in the pool, then create one
    return assignedAgent || new AdvancedAgent(extend({ persistent : true }, agent));
};

/**
 * Default status codes filter.
 * Accept or decline status code, allows or decline retries
 * @see https://github.com/nodules/asker#response-status-codes-processing
 *
 * @param {Number} code
 * @returns {{accept : Boolean, isRetryAllowed : Boolean}}
 */
Request.prototype.statusCodeFilter = function(code) {
    return {
        accept : code === 200 || code === 201,
        isRetryAllowed : 400 > code || code > 499
    };
};

/**
 * @returns {{network: (undefined|number), total: (undefined|number)}}
 */
Request.prototype.getTimers = function() {
    return {
        network : this._networkTime ? this._networkTime.time : void(0),
        total : this._executionTime ? this._executionTime.time : void(0)
    };
};

/**
 * timestamp for error messages in the '[in XX~YY ms]' format
 * XX – time between socket was assigned for request and response or error retrieved
 * YY – time about `execute` method call and response or error retrieved
 * @returns {String}
 */
Request.prototype.formatTimestamp = function() {
    var timers = this.getTimers();

    // @todo let user to use custom format
    return ['in ', timers.network || '0', '~', timers.total, ' ms'].join('');
};

/**
 * @returns {{time:{network:number,total:number},retries:{used:number,limit:number}}}
 */
Request.prototype.getResponseMetaBase = function() {
    return {
        time : this.getTimers(),
        options : this.options,
        retries : {
            used : this.retries,
            limit : this.options.maxRetries
        }
    };
};

/**
 * Getter of the private _isRunning flag
 *
 * @memberOf Request.prototype
 * @field {Boolean} isRunning
 */
Object.defineProperty(Request.prototype, 'isRunning', {
    get : function() {
        return Boolean(this._isRunning);
    },
    enumerable : true
});

/**
 * @returns {Object} used to fill common placeholders of Asker errors (%timings, %url, %requestId%)
 */
Request.prototype.getCommonErrorData = function() {
    return {
            timings : this.formatTimestamp(),
            url : this.getUrl(),
            requestId : this.options.requestId
        };
};

/**
 * @param {String} timerId Short timer name
 * @returns {String} Full timer nme, which satisfies the pattern "asker.<requestId>.<timerId>"
 */
Request.prototype.buildTimerId = function(timerId) {
    var base = 'asker.';

    if (this.options.requestId) {
        base += this.options.requestId + '.';
    }

    return base + timerId;
};

/**
 * Set _isRunning flag to false and call the callback if any
 *
 * @param {Error} err
 * @param {Object} [data]
 * @returns {*}
 */
Request.prototype.done = function(err, data) {
    this._isRunning = false;

    if ( ! this._networkTime) {
        this._networkTime = contimer.stop(this, this.buildTimerId('network'));
    }

    if ( ! this._executionTime) {
        this._executionTime = contimer.stop(this, this.buildTimerId('execution'));
    }

    if (err instanceof AskerError) {
        err.bind(this.getCommonErrorData());
    }

    if (typeof this._callback === 'function') {
        this._callback(err, data);
    }
};

/**
 * Successfully resolve request
 *
 * @param {Number} code http status code
 * @param {*} response data
 * @param {Object} [meta]
 * @param {Object} headers http response headers
 */
Request.prototype.resolve = function(code, response, meta, headers) {
    this._executionTime = contimer.stop(this, this.buildTimerId('execution'));

    this.done(null, {
        data : response,
        statusCode : code,
        headers : headers,
        meta : meta ? extend(this.getResponseMetaBase(), meta) : this.getResponseMetaBase()
    });
};

/**
 * @returns {String} request URL for errors details producing
 */
Request.prototype.getUrl = function() {
    return [
            this.options.protocol,
            '//',
            this.options.hostname,
            ':',
            this.options.port,
            this.options.path
        ].join('');
};

/**
 * Handle request retries. Throw an error if retries limit exceeded.
 * @param {Object} requestOptions same as for _tryHttpRequest
 * @param {AskerError} retryReason error which is a reason for retry
 * @private
 */
Request.prototype._retryHttpRequest = function(requestOptions, retryReason) {
    if (this.retries >= this.options.maxRetries) {
        // retries limit exceeded
        // throw an RETRIES_LIMIT_EXCEEDED if retries allowed for request
        // or retry reason error in another case
        if (this.options.maxRetries > 0) {
            this._executionTime = contimer.stop(this, this.buildTimerId('execution'));

            // @todo throw following error with `reason` prop which contains retryReason
            // so user code can determine limits exceeded errors
            retryReason = AskerError
                .createError(
                    AskerError.CODES.RETRIES_LIMIT_EXCEEDED,
                    retryReason.bind(this.getCommonErrorData()))
                .bind({
                    maxRetries : this.options.maxRetries
                });
        }

        this.done(retryReason);
    } else {
        this.retries++;

        if (typeof this._onretry === 'function') {
            // call `onretry` callback if any has been passed to the constructor in the options.onretry
            // used to notify callee about retries
            this._onretry(retryReason, this.retries);
        }

        // @todo may be in the next tick?
        this._tryHttpRequest(requestOptions);
    }
};

/**
 * run the request
 * @param {Object} options request params the same as for http.request
 * @private
 */
Request.prototype._tryHttpRequest = function(options) {
    var self = this,

        requestModule = options.protocol === 'https:' ? https : http,

        /** @type {undefined|http.ClientRequest} */
        httpRequest,

        /**
         * @type {Function}
         * @param {Error} reason
         */
        retryRequest = this._retryHttpRequest.bind(this, options);

    httpRequest = requestModule.request(options, /** @param {http.IncomingMessage} res */ function(res) {
        /** @type {{ accept : Boolean, isRetryAllowed : Boolean }} */
        var statusFilterResult,
            /** @type {Array|Buffer} */
            body = [],
            /** @type Number calculated for chunked request to boost buffers concatenation */
            bodyLength = 0;

        statusFilterResult = self.statusCodeFilter(res.statusCode);

        // if status code isn't accepted by status filter
        // then abort current request execution and
        // retry reqeust or raise the UNEXPECTED_STATUS_CODE error
        if ( ! statusFilterResult.accept) {
            var error;

            httpRequest.break();

            error = AskerError.createError(
                AskerError.CODES.UNEXPECTED_STATUS_CODE,
                {
                    statusCode : res.statusCode,
                    url : self.getUrl()
                });

            if (statusFilterResult.isRetryAllowed) {
                return retryRequest(error);
            } else {
                return self.done(error);
            }
        }

        res.on('data', function(chunk) {
            body.push(chunk);
            bodyLength += chunk.length;
        });

        res.on('end', function() {
            var encoding;

            httpRequest.clearTimeouts();

            if (httpRequest.rejected) {
                // don't try to produce response if any error,
                // like http parser error, was recieved early from http client
                // @todo may be needs to be revised in the future
                return;
            }

            encoding = res.headers['content-encoding'];
            body = Buffer.concat(body, bodyLength);

            // @todo "gzip" presence test must be improved to don't pass anything like "ggzip"
            // @todo don't miss to write the test for it
            if (encoding && encoding.toLowerCase().indexOf('gzip') > -1) {
                zlib.gunzip(body, function(error, deflatedBody) {
                    if (error) {
                        self.done(AskerError.createError(AskerError.CODES.GUNZIP_ERROR, error));
                    } else {
                        self.resolve(res.statusCode, deflatedBody, { gzip : true }, res.headers);
                    }
                });
            } else {
                self.resolve(res.statusCode, body.length > 0 ? body : null, {}, res.headers);
            }
        });
    });

    /**
     * clear queue and socket timeouts if any
     */
    httpRequest.clearTimeouts = function() {
        // stop tracking time for network operations
        self._networkTime = contimer.stop(self, self.buildTimerId('network'));

        if (httpRequest.socketTimeout) {
            clearTimeout(httpRequest.socketTimeout);
            httpRequest.socketTimeout = null;
        }

        if (httpRequest.queueTimeout) {
            clearTimeout(httpRequest.queueTimeout);
            httpRequest.queueTimeout = null;
        }
    };

    /**
     * breaks request execution and retry request if errorCode provided as retry reason
     *
     * @param {Number} errorCode Asker.Error code
     * @param {Object} errorData data for Asker.Error message interpolation
     */
    httpRequest.break = function(errorCode, errorData) {
        httpRequest.clearTimeouts();

        // mark this request as rejected, response must not be built in this case
        httpRequest.rejected = true;
        httpRequest.abort();

        // force socket removing due to errors
        httpRequest.emit('removeSocket');

        if (errorCode) {
            // call for retry if error provided as the reason for it
            retryRequest(AskerError.createError(errorCode, errorData));
        }
    };

    // setup queue timeout
    httpRequest.queueTimeout = setTimeout(function() {
        httpRequest.break(AskerError.CODES.QUEUE_TIMEOUT);
    }, options.queueTimeout);

    // socket assigned to request
    httpRequest.on('socket', function() {
        // start tracking time of the network operations
        contimer.start(self, self.buildTimerId('network'));

        httpRequest.socketTimeout = setTimeout(function() {
            httpRequest.break(AskerError.CODES.SOCKET_TIMEOUT);
        }, options.timeout);
    });

    httpRequest.on('error', function(error) {
        if ( ! httpRequest.rejected) {
            // don't try to break request execution twice or more
            httpRequest.break(AskerError.CODES.HTTP_CLIENT_REQUEST_ERROR, error);
        }
    });

    // send request body
    if (typeof this.options.body !== 'undefined') {
        httpRequest.write(this.options.body);
    }

    httpRequest.end();
};

/**
 * Execute request
 * @param {Function} [callback]
 * @returns {boolean} `false` if request already running
 */
Request.prototype.execute = function(callback) {
    // don't try to execute already running request
    if (this.isRunning) {
        return false;
    }

    // @todo may be throw error if _callback is defined already
    if (typeof callback === 'function') {
        this._callback = callback;
    }

    this.retries = 0;
    this._isRunning = true;

    // start tracing total request execution time
    // including networks ops time, asker code execution and queue in the pool
    contimer.start(this, this.buildTimerId('execution'));

    this._tryHttpRequest(extend({}, this.options, { agent : Request.getAgent(this) }));

    return true;
};

module.exports = Request;
